plugins/reflection/src/main/java/org/qbicc/plugin/reflection/ReflectionAnalyzingBasicBlockBuilder.java

```diff
@@ -0,0 +1,200 @@
+import org.qbicc.graph.Value;
+import org.qbicc.graph.literal.ExecutableLiteral;
+import org.qbicc.graph.literal.StringLiteral;
+import org.qbicc.graph.literal.TypeLiteral;
+import org.qbicc.interpreter.VmObject;
+import org.qbicc.plugin.reachability.ReachabilityRoots;
+import org.qbicc.type.ObjectType;
+import org.qbicc.type.definition.DefinedTypeDefinition;
+import org.qbicc.type.definition.LoadedTypeDefinition;
+import org.qbicc.type.definition.element.FieldElement;
+import org.qbicc.type.definition.element.MethodElement;
+
+import java.util.List;
+import java.util.Map;
+
+/**
+ * This BBB provides automatic registration/folding of reflective operations
+ * with constant arguments. Like GraalVM's "Automatic Detection"
+ * of reflection (https://www.graalvm.org/22.0/reference-manual/native-image/Reflection/),
+ * the goal is to support simple patterns for reflection without explicit user annotation.
+ */
+public class ReflectionAnalyzingBasicBlockBuilder extends DelegatingBasicBlockBuilder {
+    private static final AttachmentKey<KnownMethods> KEY = new AttachmentKey<>();
+
+
+    private final CompilationContext ctxt;
+
+    public ReflectionAnalyzingBasicBlockBuilder(final FactoryContext ctxt, final BasicBlockBuilder delegate) {
+        super(delegate);
+        this.ctxt = getContext();
+    }
+
+    @Override
+    public Value call(Value targetPtr, Value receiver, List<Value> arguments) {
+        Value replacement = process(targetPtr, receiver, arguments);
+        return replacement != null ? replacement : super.call(targetPtr, receiver, arguments);
+    }
+
+    @Override
+    public Value callNoSideEffects(Value targetPtr, Value receiver, List<Value> arguments) {
+        Value replacement = process(targetPtr, receiver, arguments);
+        return replacement != null ? replacement : super.callNoSideEffects(targetPtr, receiver, arguments);
+    }
+
+
+    @Override
+    public Value invoke(Value targetPtr, Value receiver, List<Value> arguments, BlockLabel catchLabel, BlockLabel resumeLabel, Map<Slot, Value> targetArguments) {
+        Value replacement = process(targetPtr, receiver, arguments);
+        return replacement != null ?  replacement : super.invoke(targetPtr, receiver, arguments, catchLabel, resumeLabel, targetArguments);
+    }
+
+    private Value process(Value targetPtr, Value receiver, List<Value> arguments) {
+        if (targetPtr instanceof ExecutableLiteral el) {
+            KnownMethods km = KnownMethods.get(ctxt);
+
+            // If the target is a method of java.lang.Class, then see if it is one of the reflection entry points
+            // we want to handle specially.
+            if (el.getExecutable() instanceof MethodElement me && me.getEnclosingType().load().equals(km.jlc)) {
+                // replace Class.forName(<literal String>) with a class literal if the type can be resolved at build time
+                if (me.equals(km.forName) && arguments.get(0) instanceof StringLiteral sl) {
```

Comment:
It does still seem like this would be easier (and more efficient) using the intrinsics API to register these hooks, having them return `null` if the criteria are not met. Was there a limitation or problem with the intrinsic API that pushed you in this direction?

Reply:
I was mentally conflating intrinsics and interpreter hooks when I was working through how this could work.

I also had the thought that a BBB could be run multiple times (for example in the ANALYZE phase).  However, we aren't really setup to construct any of the Method/Field/Constructor data structures we would need after the ADD phase completes so even if we did run an optimizer pass in ANALYZE, we'd still need something in ADD.

So, changed it to be via intrinsics.  The actual code isn't that different, just the setup and boilerplate around it.